workflow-ui/core-common/src/main/java/com/squareup/workflow1/ui/navigation/NavigationMonitor.kt

```diff
@@ -0,0 +1,52 @@
+package com.squareup.workflow1.ui.navigation
+
+import com.squareup.workflow1.ui.Compatible
+import com.squareup.workflow1.ui.unwrap
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.onEach
+
+/**
+ * Reports navigation across a series of calls to [update], probably made
+ * for each rendering posted by
+ * [renderWorkflowIn][com.squareup.workflow1.renderWorkflowIn].
+ *
+ * Takes advantage of [unwrap()] and [Compatible.keyFor] to provide navigation
+ * logging by reporting the top (read: last-most, inner-most) sub-rendering,
+ * which conventionally is the one that is visible and accessible to the user.
+ *
+ * Reports each time the [Compatible.keyFor] the top is unequal to the previous one,
+ * which conventionally indicates that a new view object will replace the previous one.
+ */
+public class NavigationMonitor(
+  skipFirstScreen: Boolean = false,
+  private val onNavigate: (Any) -> Unit = { println(it) }
+) {
+  @Volatile
+  private var lastKey: String? = if (skipFirstScreen) null else ""
+
+  /**
+   * Uses [unwrap] to find the topmost element of [rendering] and
+   * reports it with [onNavigate] if [Compatible.keyFor] reveals that
+   * it is of a different kind from the previous top.
+   */
+  public fun update(rendering: Any) {
+    val unwrapped = rendering.unwrap()
+
+    Compatible.keyFor(unwrapped).takeIf { it != lastKey }?.let { newKey ->
+      if (lastKey != null) onNavigate(unwrapped)
+      lastKey = newKey
```

Comment:
@steve-the-edwards this is the order of these operations in house. Wondering if I should assign `lastKey` before calling `onNavigate`:

```kotlin
lastKey.let { lastKeyOrNull ->
  lastKey = newKey
  if (lastKeyOrNull != null) onNavigate(unwrapped)
}
```

Only rationale is to better handle any terrible choices a dev makes that leads a workflow update as a side effect of `onNavigate()`, so that a recursive call won't have its `lastKey` value stomped.

On the one hand, why even pretend to support that. On the other hand, it'll be maddening to track down if it happens.

Reply:
@pyricau WDYT?

Reply:
I don't think you should aim to recover right if a dev triggers a workflow update as a side effect of `onNavigate()`. That just seems wrong?

So, two options:
- A) Don't care, if it happens all bets are off
- B) Make sure it can't happen: enforce single threadness, set a boolean field while calling onNavigate to throw if we reenter.

Reply:
I'm going to err on the side of under- rather than over-engineering.